src/services/patternMatcher.ts

```diff
@@ -0,0 +1,810 @@
+module ts {
+    // Note(cyrusn): this enum is ordered from strongest match type to weakest match type.
+    export enum PatternMatchKind {
+        Exact,
+        Prefix,
+        Substring,
+        CamelCase
+    }
+
+    // Information about a match made by the pattern matcher between a candidate and the 
+    // search pattern.
+    export interface PatternMatch {
+        // What kind of match this was.  Exact matches are better than prefix matches which are 
+        // better than substring matches which are better than CamelCase matches.
+        kind: PatternMatchKind;
+
+        // If this was a camel case match, how strong the match is.  Higher number means
+        // it was a better match.
+        camelCaseWeight: number;
+
+        // If this was a match where all consituent parts of the candidate and search pattern 
+        // matched cast sensitively or case insensitively.  Case sensitive matches of the kind
+        // are better matches than insensitive matches.
+        isCaseSensitive: boolean;
+
+        // Whether or not this match occured with the punctuation from the search pattern stripped
+        // out or not.  Matched without the punctuation stripped are better than ones with punctuation
+        // stripped.
+        punctuationStripped: boolean;
+    }
+
+    // The pattern matcher maintains an internal cache of information as it is used.  Therefore,
+    // you should not keep it around forever and should get and release the matcher appropriately

```

Comment:
Weird to heard "you" in the comment but maybe it's good because it captures the reader's attention.